SYCL/ESIMD/unused_spec_const.cpp

```diff
@@ -0,0 +1,49 @@
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// This test checks that if specialization constant is set but not used, then
+// 1) The kernel still works properly;
+// 2) The kernel parameter holding the buffer for specialization constants is
+//    optimized away.
+//    TODO: the second part of the check should be added to this test when
+//    DAE (Dead Arguments Elimination) optimization is enabled for ESIMD.
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <sycl/ext/intel/esimd.hpp>
+
+constexpr sycl::specialization_id<int> Spec;
+
+int main() {
+  sycl::queue Q;
+  std::cout << "Running on: "
+            << Q.get_device().get_info<sycl::info::device::name>() << std::endl;
+
+  int *A = sycl::malloc_shared<int>(1, Q);
+  *A = 123;
+  Q.submit([&](sycl::handler &CGH) {
+     CGH.set_specialization_constant<Spec>(42);
+     CGH.single_task([=](sycl::kernel_handler KH) SYCL_ESIMD_KERNEL {
+       // This use of get_specializaiton_constant is commented out.
+       // Thus the device code doen't have info about it and cannot allocate
+       // memory/buffer for it on device side.
+       // The test checks that the unused kernel parameter is either
+       // eliminated by DAE optimization OR nullptr is passed for that argument.
+
+       // KH.get_specialization_constant<Spec>();
+       *A += 2;
```

Comment:
formally speaking, we don't support C++ style memory operations in ESIMD (even though they sometimes work)

Reply:
wait, it is not memory operation. It is "*A = *A + 2;"

Reply:
Thank you. I replaced it with ESIMD load/store: 30ab516311043d6a989f3a03200d7048b00d5af0